src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp

```diff
@@ -929,7 +935,19 @@ HeapWord* ShenandoahHeap::allocate_memory_under_lock(ShenandoahAllocRequest& req
 
 HeapWord* ShenandoahHeap::mem_allocate(size_t size,
                                         bool*  gc_overhead_limit_was_exceeded) {
+  *gc_overhead_limit_was_exceeded = false;
   ShenandoahAllocRequest req = ShenandoahAllocRequest::for_shared(size);
+  if (get_gc_no_progress_count() > ShenandoahNoProgressThreshold) {
+    // Shenandoah will grind along for quite a while allocating one
+    // object at a time using non-tlab allocations. This will notify
+    // the collector to start a cycle, but will raise an OOME to the
+    // mutator if the last Full GCs have not made progress.
+    control_thread()->handle_alloc_failure(req, false);
+    notify_gc_progress();
+    *gc_overhead_limit_was_exceeded = true;
```

Comment:
I am confused. We have noted that we did more than `ShenandoahNoProgressThreshold` collections at this point. So we will fail the allocation request, set `gc_overhead_limit_was_exceeded`, and return `null` to the requester.

What happens at the caller, in response to this set of conditions?

Why then does it make sense to `notify_gc_progress()`, which clears the counter of `_gc_no_progress_count`? Where was the ostensible progress in this case?

Reply:
Ultimately, the caller will have an OOME raised. The `_gc_no_progress_count` is cleared to allow subsequent mutator allocations to resume the slower block-and-retry loop in `ShenandoahHeap::allocate_memory`.

Reply:
OK, thanks, that makes sense. Can you add a comment line to that effect?